### PR TITLE
Unarchiving Support

### PR DESCRIPTION
**Pull request by mitchellh**
This adds support for downloading archives with go-getter and automatically unarchiving them.

We use the magic query parameter `archive` to specify the archive type. We also have shorthand by detecting the extension and using that to populate `archive` if it isn't set. Supported archive types are: `tar.gz`, `tar.bz2`, `gz`, `bz2`, `tgz`, `tbz`, and `zip`. More types can be very easily added.

Archiving can be disabled explicitly (if you're actually trying to just download the archive itself) by setting `archive=false` (or 0, anything that ParseBool's to false). 

This will instantly enable the following in our projects:

  * Terraform modules can be archived
  * Nomad binaries can be archived
  * Otto dependencies and imports can be archived

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -105,10 +119,59 @@ func (c *Client) Get() error {
 			"download not supported for scheme '%s'", force)
 	}
 
+	// We have magic query parameters that we use to signal different features
+	q := u.Query()
+
+	// Determine if we have an archive type
+	archiveV := q.Get("archive")
+	if archiveV != "" {
+		q.Del("archive")
```
Comment (dadgar) [inline client.go]: I would put a comment on why you delete

File: client.go [hunk 2]
```diff
@@ -211,3 +300,12 @@ func checksum(source string, h hash.Hash, v []byte) error {
 
 	return nil
 }
+
+var archiveMagic = map[string]struct{}{
```
Comment (dadgar) [inline client.go]: Unused?

Comment (mitchellh) [inline client.go]: Fixed in a later commit :)

**Pull request merged**